Write how-to for Cloudscale installation (#21)
appuio/openshift4-docs: docs/modules/ROOT/pages/how-tos/install/cloudscale.adoc

> = Installation on Cloudscale

Reviewer: ```suggestion
= Installation on cloudscale.ch
```

[abstract]
--

> Steps to install an OpenShift 4 cluster on https://cloudscale.ch[Cloudscale].

Reviewer: ```suggestion
Steps to install an OpenShift 4 cluster on https://cloudscale.ch[cloudscale.ch].
```

These steps follow the https://docs.openshift.com/container-platform/latest/installing/installing_bare_metal/installing-bare-metal.html[Installing a cluster on bare metal] docs to set up a user provisioned installation (UPI).
https://www.terraform.io[Terraform] is used to provision the cloud infrastructure.
--

== Prerequisites

> * Cloudscale API token

Reviewer: ```suggestion
* cloudscale.ch API token
```

> * `terraform`

Reviewer: We might need to mention the Terraform version, either here or in the Terraform repository.

Author: It is already enforced here: https://github.com/appuio/terraform-openshift4-cloudscale/blob/init/provider.tf#L2

* `mc` https://docs.min.io/docs/minio-client-quickstart-guide.html[Minio client]
* `jq`
* Clone of the https://github.com/appuio/terraform-openshift4-cloudscale[terraform-openshift4-cloudscale] repository

== Cluster Installation

> . Create a new Syn cluster: https://control.vshn.net/syn/lieutenantclusters

Reviewer: ```suggestion
. Register the new OpenShift 4 cluster in Lieutenant: https://control.vshn.net/syn/lieutenantclusters
```

. Remove bootstrap node
+
[source,console]
----
terraform apply \
  -target=cloudscale_server.bootstrap
----

. Provision worker nodes
+
[source,console]
----
terraform apply

export KUBECONFIG=${CLUSTER_NAME}/auth/kubeconfig

# Once CSRs in state Pending show up, approve them
# Needs to be run twice
kubectl get csr
kubectl get csr --no-headers | \
  grep Pending | awk '{ print $1 }' | xargs \
  kubectl certificate approve

kubectl get nodes
----

. Configure router HAProxy
+
[source,console]
----
echo "router_servers=$(terraform output -json | jq .router_servers.value)" \
  > terraform.tfvars

terraform apply
----

. Create secret with S3 credentials https://docs.openshift.com/container-platform/4.5/registry/configuring_registry_storage/configuring-registry-storage-aws-user-infrastructure.html#registry-operator-config-resources-secret-aws_configuring-registry-storage-aws-user-infrastructure[for the registry]
+
[source,console]
----
oc create secret generic image-registry-private-configuration-user \
--namespace openshift-image-registry \
--from-literal=REGISTRY_STORAGE_S3_ACCESSKEY=${AWS_ACCESS_KEY_ID} \
--from-literal=REGISTRY_STORAGE_S3_SECRETKEY=${AWS_SECRET_ACCESS_KEY}
----

> . Synfection

Reviewer: ```suggestion
. Make the cluster Project Syn enabled
```

+

> Synfect the cluster according to https://wiki.vshn.net/x/ngMBCg

Reviewer: ```suggestion
Install Steward on the cluster according to https://wiki.vshn.net/x/ngMBCg
```

= Installation on Cloudscale

:toc:

[abstract]
.General Procedure
--
. Render install manifests

. Render ignition config

. Provision bootstrap and control plane nodes

. Provision worker and LB nodes

. Synfection

. Post install configuration
--

.Prerequisites
* Cloudscale API token
* `terraform`
* `mc` https://docs.min.io/docs/minio-client-quickstart-guide.html[Minio client]
* `jq`
* Cloned https://github.com/appuio/terraform-openshift4-cloudscale[terraform-openshift4-cloudscale] repository

== Cluster Installation

. Create a new Syn cluster: https://control.vshn.net/syn/lieutenantclusters

. Configure input
+
[source,console]
----
export CLOUDSCALE_TOKEN=<cloudscale-token> # From https://control.cloudscale.ch/user/api-tokens
export CLUSTER_NAME=<cluster-name>
export BASE_DOMAIN=ocp4-poc.appuio-beta.ch
export PULL_SECRET=<redhat-pull-secret> # From https://cloud.redhat.com/openshift/install/pull-secret
----

. Create S3 buckets
+
[source,console]
----
# Use already exiting bucket user

> curl -H "Authorization: Bearer ${CLOUDSCALE_TOKEN}" \

Reviewer 2: Do I have to do something with the output? I guess, wetter I use an existing user user or create one, the `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` variables must be set. I can not see the result of this command being used anywhere (e.g. assigned to the variable `response`).

  https://api.cloudscale.ch/v1/objects-users | \
  jq -e ".[] | select(.display_name == \"${CLUSTER_NAME}\")"

# Or create a new one
response=$(curl -H "Authorization: Bearer ${CLOUDSCALE_TOKEN}" \
  -F display_name=${CLUSTER_NAME} \
  https://api.cloudscale.ch/v1/objects-users)

export AWS_ACCESS_KEY_ID=$(echo $response | jq -r .keys[0].access_key)
export AWS_SECRET_ACCESS_KEY=$(echo $response | jq -r .keys[0].secret_key)

mc config host add \
  s3 https://objects.rma.cloudscale.ch ${AWS_ACCESS_KEY_ID} ${AWS_SECRET_ACCESS_KEY}

mc mb --ignore-existing\
  "s3/${CLUSTER_NAME}-bootstrap-ignition" \

>   "s3/${CLUSTER_NAME}-tf-state"

Reviewer 2: With AKS and EKS we use one single S3 bucked (on AWS) to store the states. We then use workspaces to switch between states of different environments.

# Ignore the error
mc policy set download "s3/${CLUSTER_NAME}-bootstrap-ignition"
----

. Prepare `install-config.yaml`
+
[source,console]
----
mkdir ${CLUSTER_NAME}

cat > "${CLUSTER_NAME}/install-config.yaml" <<EOF

> apiVersion: v1

Reviewer 2: If I am not mistaken, most of that config is default anyway. We have this now all over the place. Can we just create configs with the bare minimum? Does this come from the OpenShift documentation?

Author: Yes, this is the bare minimum required for an installation.

Reviewer 2: I remember that in last fall, we have created clusters with only a hand full of settings. Al those master, worker and networking setting could just be ommited.

  serviceNetwork:
    - 172.30.0.0/16
platform:
  none: {}
pullSecret: |
  ${PULL_SECRET}
sshKey: "$(cat ~/.ssh/id_ed25519.pub)"
EOF

----

. Prepare install manifests and ignition config
+
[source,console]
----
openshift-install --dir ${CLUSTER_NAME} \
  create manifests

openshift-install --dir ${CLUSTER_NAME} \
  create ignition-configs

mc cp ${CLUSTER_NAME}/*.ign "s3/${CLUSTER_NAME}-bootstrap-ignition"
----

. Provision infrastructure
+
[source,console]
----
terraform init \
  -backend-config "bucket=${CLUSTER_NAME}-tf-state"

export TF_VAR_cluster_name=$CLUSTER_NAME

terraform apply \
  -target=cloudscale_floating_ip.api_vip \
  -var api_eip=""

terraform refresh

export TF_VAR_api_eip="$(terraform output -json | jq -r .api_vip.value)"

terraform apply \
  -target=cloudscale_floating_ip.api_vip \
  -target=cloudscale_server.bootstrap \
  -target=cloudscale_server.master \
  -var bootstrap_count=1
----

. Create the necessary DNS records
+
[source,console]
----
terraform refresh

terraform output -json | jq -r .dns_entries.value
----

. Wait for bootstrap to complete
+
[source,console]
----
openshift-install --dir ${CLUSTER_NAME} \
  wait-for bootstrap-complete
----

> . Remove bootstrap node

Reviewer 2: Can be omitted. The next step will run that target anyway.

Author: I wanted this to be a separate step to make sure the bootstrap node is gone before adding nodes.

Reviewer 2: Fair.
